[FLINK-18537][table] Remove RowDataTypeInfo

twalthr opened the pull request:
## What is the purpose of the change

`TypeInformation` is a legacy class for the sole purpose of creating a `TypeSerializer`.
 Instances of `TypeInformation` are not required in the table ecosystem but sometimes enforced
 by interfaces of other modules (such as `org.apache.flink.api.dag.Transformation`). Therefore, we recently introduced `WrapperTypeInfo` which acts as an adapter whenever type information is required. Instances of `WrapperTypeInfo` should only be created for passing it to interfaces that require type information. The class should not be used as a replacement for a `LogicalType`. Information such as the arity of a row type, field types, field names, etc. should be derived from the `LogicalType` directly.

This PR updates the code base to not use `RowDataTypeInfo` anymore. `RowDataTypeInfo` was a composite type information with legacy types as field types. This refactoring unblocks FLINK-15803 and reduces potential bugs from type conversion. 

## Brief change log

See commit messages.

## Verifying this change

This change is already covered by existing tests.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: yes
  - The serializers: yes
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? no
  - If yes, how is the feature documented? not applicable

twalthr commented:
CC @wuchong @dawidwys
Btw after all this refactoring I was thinking about renaming `WrapperTypeInfo` to `InternalTypeInfo`. This would better match to its purpose. What do you think?

dawidwys commented on flink-table/flink-table-common/src/main/java/org/apache/flink/table/types/DataTypeQueryable.java:
```diff
@@ -0,0 +1,30 @@
+ * Indicates that a {@link DataType} can be retrieved from a class that implements this interfaces.
+ */
+@PublicEvolving
+public interface DataTypeQueryable {
```
Is this interface really necessary? Especially with `@PublicEvolving` annotation? How are users supposed to use it? If I understand it correctly you need it for internal operations. Moreover you need it because the `WrapperTypeInfo` is in `blink-runtime`, right?

Can't we move the `WrapperTypeInfo` to the `table-common` instead?  The class itself has no runtime dependencies. Only the factory methods need some runtime classes.

twalthr replied:
I thought about this as well and decided for keeping `WrapperTypeInfo` in the runtime package. The reason for this is to keep the serializers close to this utility class in the same package. `WrapperTypeInfo` is not needed in `table-common` because it is internal. Actually the constructor of `WrappenTypeInfo` should even be private but this is not possible due to other failing tests so far. We can mark `DataTypeQueryable` internal if that helps, but the past around `TypeInformation` has shown that you need such a simple interface over and over again. It will be more useful in the future once we introduce a "converter type serializer" where the return data type is not of internal data structure.

wuchong replied:
I prefer to keep `WrapperTypeInfo` in `blink-runtime`, especially we have a public construct which will cause a lot of problems if the `logicalType` and `typeSerializer` are not mapped.

wuchong replied:
Is it possible to make the construct of `WrapperTypeInfo` to be package visible? We can move the utility `TypeInfoDataTypeConverter#createWrapperTypeInfo` into `WrapperTypeInfo`. And the remaining constructor invoking are the tests under the same package.

wuchong commented on flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/data/util/DataFormatConverters.java:
```diff
@@ -249,7 +251,15 @@ public static DataFormatConverter getConverterForDataType(DataType originDataTyp
+				TypeInformation<?> asTypeInfo = fromDataTypeToTypeInfo(dataType);
+				if (asTypeInfo instanceof WrapperTypeInfo && clazz == RowData.class) {
+					LogicalType realLogicalType = ((WrapperTypeInfo<?>) asTypeInfo).toLogicalType();
+					return new RowDataConverter(getFieldCount(realLogicalType));
```
In which case will we reach here?

twalthr replied:
All cases that went through `TypeInfoDataTypeConverter#fromDataTypeToTypeInfo` e.g. the legacy sinks.

wuchong commented on flink-table/flink-table-runtime-blink/src/test/java/org/apache/flink/table/runtime/operators/join/AsyncLookupJoinHarnessTest.java:
```diff
@@ -75,18 +74,19 @@
-		Types.INT,
-		Types.STRING
+	private final RowDataHarnessAssertor assertor = new RowDataHarnessAssertor(new LogicalType[]{
+		new IntType(),
+		new VarCharType(VarCharType.MAX_LENGTH),
+		new IntType(),
+		new VarCharType(VarCharType.MAX_LENGTH)
```
Keep to use DataType here? 

```java
                DataTypes.INT().getLogicalType(),
		DataTypes.STRING().getLogicalType(),
		DataTypes.INT().getLogicalType(),
		DataTypes.STRING().getLogicalType()
```

twalthr commented on flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/codegen/CodeGeneratorContext.scala:
```diff
@@ -18,23 +18,23 @@
 
 package org.apache.flink.table.planner.codegen
 
+import java.util.TimeZone
+
+import org.apache.calcite.avatica.util.DateTimeUtils
```
Thanks for the reminder. I only set this for Java not for Scala. However, it difficult to enforce this without a checkstyle plugin in place.

twalthr commented:
Thanks for the feedback @wuchong and @dawidwys. I hope I addressed all comments. I renamed `WrapperTypeInfo` to `InternalTypeInfo` and introduced `ExternalTypeInfo` already for the raw type and for the future bridging to type information of DataStream API.
